cfgrib/messages.py

```diff
@@ -318,43 +308,38 @@ def __len__(self) -> int:
         return sum(1 for _ in self.items())
 
 
-@contextlib.contextmanager
-def compat_create_exclusive(path):
-    # type: (str) -> T.Generator[T.IO[bytes], None, None]
-    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL)
-    with open(fd, mode="wb") as file:
-        try:
-            yield file
-        except Exception:
-            file.close()
-            os.unlink(path)
-            raise
+ALLOWED_PROTOCOL_VERSION = "1"
 
 
-ALLOWED_PROTOCOL_VERSION = "1"
+C = T.TypeVar("C", bound="FieldsetIndex")
 
 
 @attr.attrs(auto_attribs=True)
-class FileIndex(abc.Index[OffsetType, Message]):
-    filestream: FileStream
+class FieldsetIndex(abc.Index[T.Any, abc.Field]):
+    fieldset: abc.Fieldset[T.Any, abc.Field]
     index_keys: T.List[str]
-    offsets: OffsetsType = attr.attrib(repr=False)
     filter_by_keys: T.Dict[str, T.Any] = {}
+    field_ids_index: T.List[T.Tuple[T.Tuple[T.Any, ...], T.List[abc.Field]]] = attr.attrib(
+        repr=False, default=[]
+    )
     index_protocol_version: str = ALLOWED_PROTOCOL_VERSION
 
     @classmethod
-    def from_filestream(cls, filestream, index_keys):
-        # type: (FileStream, T.Sequence[str]) -> FileIndex
-        offsets = {}  # type: T.Dict[T.Tuple[T.Any, ...], T.List[OffsetType]]
+    def from_fieldset(cls, fieldset, index_keys):
+        # type: (T.Type[C], abc.Fieldset[T.Any, abc.Field], T.Sequence[str]) -> C
+        field_ids_index = {}  # type: T.Dict[T.Tuple[T.Any, ...], T.List[T.Any]]
         index_keys = list(index_keys)
-        count_offsets = {}  # type: T.Dict[int, int]
         header_values_cache = {}  # type: T.Dict[T.Tuple[T.Any, type], T.Any]
-        for offset_field, message in filestream.items():
+        if hasattr(fieldset, 'items') and callable(fieldset.items):
+            iterate = fieldset.items()
+        else:
+            iterate = enumerate(fieldset)
```

Comment:
@b8raoult `mypy` gets confused because the two assign to `itarate` are not the same type. I'll fix it.